feat: add mpool updates

frrist opened the pull request:
Part of https://github.com/filecoin-project/sentinel/issues/33

frrist committed:
misc: add .idea to .gitignore

frrist commented on plugins/inputs/lotus/lotus.go:
```diff
@@ -296,13 +335,38 @@ func recordTipsetStatePoints(ctx context.Context, api api.FullNode, acc telegraf
+		},
+		map[string]string{
+			"message_cid_tag":       newMpoolUpdate.Message.Cid().String(),
+			"mpool_update_type_tag": updateType,
```
This wasn't specified in the acceptance criteria: Happy to remove if unwanted. Its value indicates if the message is being added to the pool or removed from the pool.

placer14 replied:
I was going to ask for it to be removed as I believe it's redundant to information we capture with chainwatch (since we can assume a msg is removed from the pool when it's added to a block), but I think we want the datasets to be independent in case we come across a sentinel configuration which does not have chainwatch included. Thanks for the consideration.

frrist replied:
Does that mean keep it or remove it?

It's also worth noting that messages can be removed (what lotus calls "pruning") from the pool for reasons other than being included in a block, e.g. invalid fields, pool being too full, insufficient gas, etc. I think keeping a record of this adds value.

placer14 replied:
Sorry, yes keep it.

placer14 commented on plugins/inputs/lotus/lotus.go:
```diff
@@ -210,8 +239,8 @@ func processTipset(ctx context.Context, node api.FullNode, acc telegraf.Accumula
 
-func processHeader(ctx context.Context, acc telegraf.Accumulator, newHeader *types.BlockHeader, receivedAt time.Time) {
-	err := recordBlockHeaderPoints(ctx, acc, receivedAt, newHeader)
+func processHeader(ctx context.Context, acc telegraf.Accumulator, newHeader *types.BlockHeader, receivedAt time.Time, receivedFrom peer.ID) {
```
`ReceivedFrom` in this context (telegraf's perception of the node) is ambiguous. (Either "the node which you received the block from" or as you intended here "the node behind telegraf which is reporting the block which was recieved".) Maybe we can avoid the ambiguity altogether with something like `seenBy` which allows the next dev to better intuit which value is the semanticly-desired value for the context.

frrist replied:
How does `RecordedFrom` sound? As in the node we recorded the information from.

placer14 replied:
I prefer `By` so the from/to semantics don't get in the way. Maybe `RecordedByPeerID` so we label the data with `PeerID` as well, instead of having yet another CID in our schema.

placer14 commented on plugins/inputs/lotus/lotus.go:
```diff
@@ -182,21 +197,35 @@ func (l *lotus) Start(acc telegraf.Accumulator) error {
 }
 
-func processTipset(ctx context.Context, node api.FullNode, acc telegraf.Accumulator, newTipSet *types.TipSet, receivedAt time.Time) {
+func processTipset(ctx context.Context, node api.FullNode, acc telegraf.Accumulator, newTipSet *types.TipSet, receivedAt time.Time, receivedFrom peer.ID) {
```
Re: `ReceivedFrom` note. While the ambiguity is less an issue with usage in tipsets, I would prefer consist naming throughout.

placer14 commented on plugins/inputs/lotus/lotus.go:
```diff
@@ -220,11 +249,20 @@ func processHeader(ctx context.Context, acc telegraf.Accumulator, newHeader *typ
+		acc.AddError(fmt.Errorf("recording mpool update (type: %d, cid: %s): %v", newMpoolUpdate.Type, newMpoolUpdate.Message.Cid(), err))
+		return
+	}
+	log.Println("I! Processed mpool update")
```
This will be too noisy for little value. Please remove.

placer14 commented on plugins/inputs/lotus/lotus.go:
```diff
@@ -296,13 +335,38 @@ func recordTipsetStatePoints(ctx context.Context, api api.FullNode, acc telegraf
+
+	acc.AddFields("chain_mpool",
+		map[string]interface{}{
+			"message_size": newMpoolUpdate.Message.Size(),
```
Good idea. 🍻

placer14 commented on plugins/inputs/lotus/lotus.go:
```diff
@@ -296,13 +335,38 @@ func recordTipsetStatePoints(ctx context.Context, api api.FullNode, acc telegraf
+		map[string]string{
+			"message_cid_tag":       newMpoolUpdate.Message.Cid().String(),
+			"mpool_update_type_tag": updateType,
+			"recorded_from":         receivedFrom.String(),
```
Depending on what we decide for `recievedFrom` we might use that same term here. (Whatever label we use should work in this context as well.)

placer14 commented on plugins/inputs/lotus/lotus.go:
```diff
@@ -315,7 +379,9 @@ func recordTipsetMessagesPoints(ctx context.Context, api api.FullNode, acc teleg
 		},
-		map[string]string{}, ts)
+		map[string]string{
+			"recorded_from": receivedFrom.String(),
```
Same.
> Depending on what we decide for recievedFrom we might use that same term here. (Whatever label we use should work in this context as well.)

placer14 reviewed (approved):
Very miner nits. Looks great. 🍻

frrist commented on plugins/inputs/lotus/lotus.go:
```diff
@@ -26,6 +29,12 @@ const (
+const (
+	MpoolAdd       = api.MpoolAdd
+	MpoolRemove    = api.MpoolRemove
+	MpoolBootstrap = MpoolRemove + 1
```
Instead of adding a new column as https://github.com/filecoin-project/sentinel/issues/36 suggests I have extended the update types here, this alright with you @placer14?

placer14 replied:
That sounds good to me. Maybe Bootstrap shouldn't be `+1` though in case another type gets added to that iota in lotus. How about `-1`?

placer14 commented on plugins/inputs/lotus/lotus.go:
```diff
@@ -296,13 +344,54 @@ func recordTipsetStatePoints(ctx context.Context, api api.FullNode, acc telegraf
+	case MpoolRemove:
+		updateType = "MpoolRemove"
+	case MpoolBootstrap:
+		updateType = "MpoolBootstrap"
```
Another thought: These strings are a little verbose to be repeated data within that schema. How about `add`,`rm`,`init`?

frrist committed:
feat: subscribe to mpool updates & collect pending

- closes filecoin-project/sentinel#36
- makes progress towards filecoin-project/sentinel#33
